Explore: UI change for log row details

ivanahuckova opened the pull request:
**About this PR**

[Design doc ](https://docs.google.com/document/u/2/d/1loK1ItubaQQBibQPuM1_blxF5jMk1gvY8Z8UDNQzH4c)
[Figma design](https://www.figma.com/file/iQYVTrFSPOtUCPVvxVojUM/Log-message-details)

This PR is the UI part/first step of https://github.com/grafana/grafana/issues/17831. It creates log details table for each log with parsed fields and labels. The details table also contains action icons for filtering, filtering-out and seeing stats. The new UI currently works for Elastic, Loki and Influx logs. For Loki, it shows both - labels (with filter/filter out functionality) and parsed fields (with show stats functionality). For Influx and Elastic, it shows only parsed fields (with show stats functionality). 

The second part of https://github.com/grafana/grafana/issues/17831  (logs data processing, structure of logs, labels, parsed fields, metrics, ...) will be solved in the separate PR. 

Loki (dark mode):
![ezgif com-video-to-gif (13)](https://user-images.githubusercontent.com/30407135/68546207-7939c280-03d4-11ea-849e-40a3a7681278.gif)

Elastic (light mode):
![ezgif com-video-to-gif (14)](https://user-images.githubusercontent.com/30407135/68546304-0da42500-03d5-11ea-9552-3cd87357b084.gif)

ivanahuckova committed:
WIP: UI changes for log details

ivanahuckova committed:
WIP: Details table layout

ivanahuckova committed:
WIP: Fix styling, add mock data templates

ivanahuckova committed:
WIP: Add hover state

ivanahuckova committed:
WIP: Add labels to label details table

ivanahuckova committed:
WIP: Add filtering and filtering out for loki logs

ivanahuckova committed:
Explore: Add parsed fields to log details

ivanahuckova committed:
Refactor, split components

ivanahuckova committed:
Explore: Add stats panel to details table

ivanahuckova committed:
Explore: Remove highlighting and stats showing from log message

ivanahuckova commented:
@davkal @aocenas 
As mentioned above, this is WIP and this is connected only to UI/UX area. I still need to do research regarding how we want to process logs data from different data sources. I was thinking about meeting on monday/tuesday to discuss this. In the meantime, could you please look at this PR and share your thoughts.

torkelo commented on packages/grafana-ui/src/components/Logs/LogRow.tsx:
```diff
@@ -59,58 +69,72 @@ class UnThemedLogRow extends PureComponent<Props, State> {
+                  {row.timeLocal}
+                </div>
+              )}
+              <div className={cx([style.logsRowCell])}>
```
Just a small tip that your using the cx utility function everywhere:
1)  first this function is not needed as it is only for composing multiple styles together (and for composing legacy classnames plus emotion styles)
2) don't think it requires the array syntax either 
3) So int the above case youy can just write className={style.logRowCell}

ivanahuckova replied:
Yes, you are right. 🙂Updated! Thanks for the feedback.

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivanah/explore/log-rows-details

ivanahuckova committed:
Explore: Remove cx when not needed

ivanahuckova committed:
Explore: Style log stats message

ivanahuckova committed:
Explore: Clean up

ivanahuckova committed:
Explore: Refactor

ivanahuckova commented:
I've worked a little on this over the weekend: 
1. I've styled stats message according to the design. 
2. Tided up it a little, updated namings, remove console logs. 
3. Changed arrow functions to functions where arrow function is not needed. 
4. Removed cx, when not needed to add classname.

See new gifs with latest UI:
![ezgif com-video-to-gif (7)](https://user-images.githubusercontent.com/30407135/67639609-4088ec00-f8f2-11e9-82e5-b5236a555385.gif)
![ezgif com-video-to-gif (8)](https://user-images.githubusercontent.com/30407135/67639612-4383dc80-f8f2-11e9-8fa9-55ad25db75b0.gif)
![ezgif com-video-to-gif (9)](https://user-images.githubusercontent.com/30407135/67639615-44b50980-f8f2-11e9-9885-1acf57140f26.gif)

ivanahuckova committed:
Explore: Simplify LogRow

aocenas commented on packages/grafana-ui/src/components/Logs/LogDetailsRow.tsx:
```diff
@@ -0,0 +1,134 @@
+  onClickFilterLabel?: (key: string, value: string) => void;
+  onClickFilterOutLabel?: (key: string, value: string) => void;
+  createStatsForLogs?: (field: string) => void;
+  getRows?: () => LogRowModel[];
```
This is taking lots of optional props which if I understand are optional only in the sense that they are not needed if `canShowMetrics = false` but are needed if `canShowMetrics = true`. This is then a bit hard to understand from the typing, what is needed when as it is not possible to define it in the types itself easily. One way to refactor this would be to have something like this:
```
<LogDetailsRow {...props} renderOnStats={() => <LogLabelStats {...statsProps}/>}/>
```

I think that way we would have 2 components with clear props requirements while `LogDetailsRow` would have only one optional prop `renderOnStats` which would decide on whether any stats can be shown and would be responsible to render them.

ivanahuckova replied:
Yes, I totally agree. I wanted to kind of unify the LogDetailsRow, but it resulted in hard-to-understand and lots-of-optional-props component. So I will split this up and follow your advice with ```<LogDetailsRow {...props} renderOnStats={() => <LogLabelStats {...statsProp```
@torkelo this is why parser is optional -> it is always there and used for the cases, when I used LogDetailsRow with parsed fields. But with labels, we don't use it and it is not required and not used. As mentioned above, I will refactor this and split it into 2 components that have clear props requirements and clear responsibilities.

aocenas commented on public/app/plugins/datasource/loki/datasource.ts:
```diff
@@ -265,6 +265,10 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
         selector = addLabelToSelector(selector, action.key, action.value);
         break;
       }
+      case 'ADD_FILTER_OUT': {
```
What does this do if we already have that label `{label="value"}` will it add `{label="value", label!="value"}`?

ivanahuckova replied:
Yes, that is exactly what it currently does. I have checked elastic's kibana and they have the same functionality -> so it results in no logs, because there are no logs that have and doesn't have specific value for label at the same time.

But I think it would make sense to add there a logic -> remove the contradicting = or != conditionality.

davkal replied:
I think that's fine for now. Should really be up for the datasource to figure out what to do with conflicting user input.

torkelo commented on packages/grafana-ui/src/components/Logs/LogDetailsRow.tsx:
```diff
@@ -0,0 +1,134 @@
+      const allRows = getRows();
+
+      // Build value-agnostic row matcher based on the field label
+      const fieldLabel = parser!.getLabelFromField(field);
```
is parser optional? All the these ! operators make me feel that something is not correctly typed or checked.

ivanahuckova replied:
See the comment above. I will refactor LogDetailsRow and split it into 2 components that have clear props requirements and clear responsibilities. In that case, parser will be required and not optional. I will also look more closely on ! operators and see what's going on. I will keep you updated!

davkal replied:
Parsers are found dynamically per log row. If it returned one, then parsed fields are shown, which will have a button to calc the stats for the parsed field. If no parser is found, there are no parsed fields and this function cannot be triggered. Bit unclear how to model this in TS.

torkelo commented:
Think we need to start merging LogsModel with DataFrame so we can view LogDetails as just viewing all data frame fields. 

Maybe not as part of this PR but something soon. We might still need a Log model abstraction but internally it would just wrap a DataFrame. 

For example for elasticsearch, it does not look like the LogDetails is not showing other fields in the ES document but just parsing the message field.

ivanahuckova commented:
@torkelo Yes exactly! That is the plan (see[ design doc](**https://docs.google.com/document/d/1loK1ItubaQQBibQPuM1_blxF5jMk1gvY8Z8UDNQzH4c/edit#**)). I've (after consultation with Andrej) decided to split this project into 2 PRs. This one focuses strictly on UI. 
We also want to focus on logs data processing/structure. However, this will be a separate PR. I have started to look into that as well, and I was thinking about having a sync/brainstorm probably on Wednesday where we could talk about it. I would definitely like you and @marefr on this call. I will try to share my ideas in the design docs today/tomorrow, so you have time to review them and share your thoughts. 🙂

So as soon as I have better idea/more thoughts on this, I will schedule a call.

davkal commented on packages/grafana-ui/src/components/Logs/LogRow.tsx:
```diff
@@ -59,57 +71,65 @@ class UnThemedLogRow extends PureComponent<Props, State> {
-          <div className={cx([style.logsRowLocalTime])} title={`${row.timeUtc} (${row.timeFromNow})`}>
-            {row.timeLocal}
+        <div className={style.logsRowLevel} />
+        {
```
Not needed.

ivanahuckova replied:
You are right - removed!

davkal commented on packages/grafana-ui/src/components/Logs/LogRow.tsx:
```diff
@@ -59,57 +71,65 @@ class UnThemedLogRow extends PureComponent<Props, State> {
-          <div className={cx([style.logsRowLabels])}>
-            <LogLabels
+        }
+        <div>
```
Outer div is not needed.

ivanahuckova replied:
Actually it is needed, cause it creates css table-cell that includes 2 rows -> row with time and message and row with logs details.

davkal commented on packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:
```diff
@@ -235,38 +105,15 @@ class UnThemedLogRowMessage extends PureComponent<Props, State> {
-              </div>
-            )}
+            <Highlighter
+              style={{ whiteSpace: 'pre-wrap' }}
```
This style should be extracted to a constant.

ivanahuckova replied:
👍done!

davkal commented on packages/grafana-ui/src/components/Logs/LogRows.tsx:
```diff
@@ -1,12 +1,13 @@
 import React, { PureComponent } from 'react';
-import { cx } from 'emotion';
+import memoizeOne from 'memoize-one';
```
Where is this used?

ivanahuckova replied:
I have just reordered imports. It is used here:
```  makeGetRows = memoizeOne((processedRows: LogRowModel[]) => {
    return () => processedRows;
  });``

davkal reviewed (commented):
Really loving the experience already 🚀 

One thing missing now is the ability to get stats on labels, which used to work:
![Screenshot 2019-10-28 at 11 28 19](https://user-images.githubusercontent.com/859729/67675887-38738f80-f978-11e9-8e80-c600e5d35307.png)

Would also be good to put a headline above the parsed fields, which could just say "Parsed fields". We'll need another section with derived fields soon anyway.

ivanahuckova committed:
Explore: Refactor, include provided feedback

ivanahuckova committed:
Explore: Refactor, add heading for parsed fields

ivanahuckova committed:
Explore: Refactor

ivanahuckova commented:
> One thing missing now is the ability to get stats on labels, which used to work:

Fixed -> see in updated version. 
![image](https://user-images.githubusercontent.com/30407135/67716967-1bb57700-f9cd-11e9-9568-8db3354ea8d8.png)

> Would also be good to put a headline above the parsed fields, which could just say "Parsed fields".

![image](https://user-images.githubusercontent.com/30407135/67716877-e90b7e80-f9cc-11e9-9b3e-876a00561ef8.png)

davkal commented on packages/grafana-ui/src/components/Logs/LogDetailsRow.tsx:
```diff
@@ -69,43 +72,42 @@ class UnThemedLogDetailsRow extends PureComponent<Props, State> {
+  createStatsForLabels() {
+    const { getRows, parser, parsedKey, parsedValue } = this.props;
+    // If is filterable, then isLabel = true
+    const isLabel = this.props.onClickFilterLabel;
```
Can we make `isLabel` a prop so that the parent has to determine this? Then we can also use the prop to send it to the stats component to adapt the stats headline to be `M of N rows have that field` instead of `... have that label`

ivanahuckova replied:
👍absolutely! Great idea with the adaptive headline.

davkal commented on packages/grafana-ui/src/components/Logs/LogRow.tsx:
```diff
@@ -92,11 +92,9 @@ class UnThemedLogRow extends PureComponent<Props, State> {
-            <i className={showDetails ? 'fa fa-chevron-up' : 'fa fa-chevron-down'} />
-          </div>
-        }
+        <div onClick={this.toggleDetails} className={style.logsRowToggleDetails}>
```
Can we add a `title` here so it shows a tooltip to help with discovery of this feature?

ivanahuckova replied:
done

davkal reviewed (commented):
this is shaping up nicely! 🌻

ivanahuckova committed:
WIP: Include feedback, change hardcoded style values to variables

ivanahuckova committed:
Explore: Remove details information from dashboard logs panel

ivanahuckova committed:
Explore: Update heading

ivanahuckova commented:
@davkal @aocenas I have incorporated your feedback, changed hard-coded style values to variables, removed possibility to see log details from dashboard logs panel. I have also added Log Lables heading to label values, as it is visually more consistent. 

![image](https://user-images.githubusercontent.com/30407135/67875929-66530280-fb37-11e9-9637-edac901c9d60.png)

ivanahuckova committed:
Explore: Add LogDetails tests

ivanahuckova committed:
Explore: Update LogDetails tests

ivanahuckova committed:
Explore: Add LogDetailsRow tests, refactor LogDetails tests

aocenas commented on packages/grafana-ui/src/components/Logs/LogDetails.tsx:
```diff
@@ -0,0 +1,107 @@
+    parsedFields: [],
+  };
+
+  parseMessage() {
```
hmm I would probably memoize this and do it during render, otherwise you should be doing this also if row.entry was updated from new props.

ivanahuckova committed:
Explore: Memoize parseMessage

ivanahuckova committed:
Merge: Merge test branch

ivanahuckova committed:
Merge branch 'master' into ivanah/explore/log-rows-details

ivanahuckova committed:
Merge branch 'ivanah/explore/log-rows-details' of https://github.com/grafana/grafana into ivanah/explore/log-rows-details

ivanahuckova committed:
Explore: Fix style change

ivanahuckova committed:
Explore: Merge conflicts resolved

aocenas commented:
Seems like with elastic we lost the field labels:
<img width="889" alt="Screenshot 2019-10-31 at 19 12 23" src="https://user-images.githubusercontent.com/1014802/67974274-7095ff00-fc12-11e9-8ce2-d9628c3f3e3d.png">

The document json is also duplicated in the message for some reason.

ivanahuckova committed:
Explore: Move parsing of message to row

ivanahuckova committed:
Explore: Remove unused imports

ivanahuckova committed:
Explore: Fix style

aocenas committed:
Fix field matchers

aocenas committed:
Move parsing back into LogDetails

corford commented:
@ivanahuckova These UI changes are great (we use the details view all the time in explore). Are there any plans to make this same details view/toggle available for Log panels in dashboards? (we've tried setting `allowDetails: true` in the panel config but it has no effect.

ivanahuckova commented:
@corford I am very happy to hear that! ❤️Regarding your question, there is currently no github issue for that. So if log details view in logs panel would be useful for you, it would be awesome if you've created the github issue for that and we will add it to our [backlog]:https://github.com/grafana/grafana/issues/20449. 🙂

corford commented:
Thanks @ivanahuckova  I've created the feature request here: https://github.com/grafana/grafana/issues/22271
